tests/Functional/Action/Lti/UpdateLti1p3OutcomeActionTest.php

```diff
@@ -0,0 +1,241 @@
+        $this->assertStringContainsString(
+            'JWT access token scopes are invalid',
+            (string)$this->kernelBrowser->getResponse()->getContent()
+        );
+    }
+
+    public function testItReturnsUnauthorizedResponseWithoutBearer(): void
+    {
+        $this->kernelBrowser->request(
+            'POST',
+            '/api/v1/lti1p3/outcome',
+            [],
+            [],
+            [
+                'CONTENT_TYPE' => 'text/xml',
+            ],
+            ''
+        );
+
+        $response = $this->kernelBrowser->getResponse();
+        $this->assertInstanceOf(Response::class, $response);
+        $this->assertEquals(Response::HTTP_UNAUTHORIZED, $response->getStatusCode());
+        $this->assertStringContainsString(
+            'A Token was not found in the TokenStorage',
+            (string)$response->getContent()
+        );
+    }
+
+    public function testItReturnsUnauthorizedResponseWithInvalidToken(): void
+    {
+        $this->kernelBrowser->request(
+            'POST',
+            '/api/v1/lti1p3/outcome',
+            [],
+            [],
+            [
+                'CONTENT_TYPE' => 'text/xml',
+                'HTTP_AUTHORIZATION' => 'Bearer invalid'
+            ],
+            ''
+        );
+
+        $response = $this->kernelBrowser->getResponse();
+        $this->assertInstanceOf(Response::class, $response);
+        $this->assertEquals(Response::HTTP_UNAUTHORIZED, $response->getStatusCode());
+        $this->assertStringContainsString('The JWT string must have two dots', (string)$response->getContent());
+    }
+
+    public function testItReturns400IfTheAuthenticationWorksButTheXmlIsInvalid(): void
+    {
+        $credentials = $this->createTestClientAccessToken(
+            $this->registration,
+            ['https://purl.imsglobal.org/spec/lti-bo/scope/basicoutcome']
+        );
+
+        $this->kernelBrowser->request(
+            'POST',
+            '/api/v1/lti1p3/outcome',
+            [],
+            [],
+            [
+                'CONTENT_TYPE' => 'text/xml',
+                'HTTP_AUTHORIZATION' => sprintf('Bearer %s', $credentials)
+            ],
+            'invalidXml'
+        );
+
+        self::assertEquals(Response::HTTP_BAD_REQUEST, $this->kernelBrowser->getResponse()->getStatusCode());
+        self::assertEquals(Assignment::STATE_READY, $this->getAssignment()->getState());
+    }
+
+    public function testItReturns404IfTheAuthenticationWorksButTheAssignmentDoesNotExist(): void
+    {
+        $credentials = $this->createTestClientAccessToken(
+            $this->registration,
+            ['https://purl.imsglobal.org/spec/lti-bo/scope/basicoutcome']
+        );
+
+        /** @var string $xmlBody */
+        $xmlBody = file_get_contents(
+            __DIR__ . '/../../../Resources/LtiOutcome/invalid_replace_result_body_wrong_assignment.xml'
+        );
```

Comment:
Let's try to avoid using such resource files if possible.

Reply:
I believe we can try to find a better solution for all these in the future, but these xml strings will need to live somewhere, if not in files, probably in the test/traits.
Once we decide the best way to handle this, we can update all tests using these files.

Reply:
Same answer as above